### ISSUE TITLE
Fail if destroy first_control_plane?

### ISSUE THREAD

**Issue by mysticaltech** [kube-hetzner/terraform-hcloud-kube-hetzner#913]
### Discussed in https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/discussions/898

<div type='discussions-op-text'>

<sup>Originally posted by **jidckii** June 21, 2023</sup>
### Description

I started this discussion here
https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/discussions/794
But it seems it deserves separate issues.

And so, I have a HA cluster and I need to update image.
But I can't do it for the first_control_plane because a lot of resources that are created inside the cluster are applied from the first instance and as a result terraform apply fails.

```
module.kube-hetzner.null_resource.kustomization: Provisioning with 'file'...
module.kube-hetzner.null_resource.kustomization: Provisioning with 'file'...
module.kube-hetzner.null_resource.kustomization: Provisioning with 'remote-exec'...
...
module.kube-hetzner.null_resource.kustomization (remote-exec): (output suppressed due to sensitive value in config)
╷
│ Error: remote-exec provisioner error
│ 
│   with module.kube-hetzner.null_resource.kustomization,
│   on .terraform/modules/kube-hetzner/init.tf line 275, in resource "null_resource" "kustomization":
│  275:   provisioner "remote-exec" {
│ 
│ error executing "/tmp/terraform_210953451.sh": Process exited with status 23

```

#### how to play

Create a cluster with
control-plane-fsn1
control-plane-nbg1
control-plane-hel1

then add
control-plane-fsn2
control-plane-nbg2
control-plane-hel2

then drain
control-plane-fsn1
control-plane-nbg1
control-plane-hel1
and set them to count = 0 in Kube.tf to remove servers.

run `terraform apply -auto-approve`

You will see the error above

### Kube.tf file

```terraform
# export TF_VAR_HCLOUD_TOKEN=${HCLOUD_TOKEN}
variable "HCLOUD_TOKEN" {
  type        = string
  description = "Token for hetzner cloud"
}

module "kube-hetzner" {
  source  = "kube-hetzner/kube-hetzner/hcloud"
  version = "2.2.3"

  providers = {
    hcloud = hcloud
  }

  depends_on = [
    tls_private_key.admin
  ]

  hcloud_token = var.HCLOUD_TOKEN

  ssh_public_key  = tls_private_key.admin.public_key_openssh
  ssh_private_key = tls_private_key.admin.private_key_openssh
  network_region  = "eu-central"

  use_control_plane_lb = true

  control_plane_nodepools = [
    {
      name        = "control-plane-fsn1",
      server_type = "cpx11",
      location    = "fsn1",
      labels      = [],
      taints      = [],
      count       = 0
    },
    {
      name        = "control-plane-nbg1",
      server_type = "cpx11",
      location    = "nbg1",
      labels      = [],
      taints      = [],
      count       = 0
    },
    {
      name        = "control-plane-hel1",
      server_type = "cpx11",
      location    = "hel1",
      labels      = [],
      taints      = [],
      count       = 0
    },

    {
      name        = "control-plane-fsn2",
      server_type = "cpx11",
      location    = "fsn1",
      labels      = [],
      taints      = [],
      count       = 1
    },
    {
      name        = "control-plane-nbg2",
      server_type = "cpx11",
      location    = "nbg1",
      labels      = [],
      taints      = [],
      count       = 1
    },
    {
      name        = "control-plane-hel2",
      server_type = "cpx11",
      location    = "hel1",
      labels      = [],
      taints      = [],
      count       = 1
    }
  ]

  agent_nodepools = [
    {
      name        = "agent-normal",
      server_type = "cpx11",
      location    = "fsn1",
      labels = [
        "nodepool=normal",
      ],
      taints = [],
      count  = 1
    }
  ]

  load_balancer_type     = "lb11"
  load_balancer_location = "fsn1"

  ingress_controller      = "none"
  enable_klipper_metal_lb = false
  enable_cert_manager     = false

  cluster_name = "test-kube-hetzner"

  create_kubeconfig    = true
  create_kustomization = true

  restrict_outbound_traffic = false

  kured_options = {
    "reboot-days" : "su,mo,tu,we,th,fr,sa"
    "start-time" : "00:00:00"
    "end-time" : "23:59:59"
    "time-zone" : "Europe/Istanbul"
  }
}


resource "tls_private_key" "admin" {
  algorithm   = "ECDSA"
  ecdsa_curve = "P384"
}

resource "local_file" "ssh_private_key" {
  content         = tls_private_key.admin.private_key_openssh
  filename        = "${path.module}/out_files/admin_key"
  file_permission = 600
}

resource "local_file" "ssh_public_key" {
  content         = tls_private_key.admin.public_key_openssh
  filename        = "${path.module}/out_files/admin_key.pub"
  file_permission = 600
}

output "kubeconfig" {
  value     = module.kube-hetzner.kubeconfig_file
  sensitive = true
}
```

### Platform

Linux</div>

**Comment by mysticaltech** [comment]
@jidckii  Fix for https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/discussions/898#discussioncomment-6585649, is coming here. Max 48h, I will keep the kustomization backup, but implement it differently so it does not depend on the first control plane. 

Thanks again for insisting on this, it's super important that the cluster survives and can be rebuilt if this node goes down.

**Comment by mysticaltech** [comment]
@jidckii See the PR above, this does fix it.

The way I tested is:

1. Create an HA cluster with 3 CP, 2 Agents.
1. Fetch the IP of the 1st CP via cat'ing the kubeconfig, match it with the server name via `hcloud server list`.
1. Issue `hcloud server delete <1st-cp-server-name>`, this simulates a catastrophic take down of the node.
1. Edit the kubeconfig with the IP of any of the other CPs.
1. Terraform apply again to restaure full cluster integrity.

During the whole process the cluster stays running perfectly.

Here's a video of the test.


https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/assets/518555/9b2b3652-0c30-4eaa-a610-b93baa850800


And the end-result after re-apply.

![ksnip_20230801-070548](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/assets/518555/cdbc18c5-2287-4f4a-bde5-c42d70b8a398)

Will merge the PR tomorrow to give time for the rest of the team to review.

**Comment by mysticaltech** [comment]
In the near future will will tag the nodes better, to know which is which more easily. This will be addressed in #903.

**Comment by jidckii** [comment]
Looks great, would you like me to test my side?

**Comment by mysticaltech** [comment]
@jidckii Of course, please do.

**Comment by mysticaltech** [comment]
Shipped in 2.5.0. Thanks again for pointing out there was an issue with that @jidckii! Very much appreciated 🙏
